ledger/src/blockstore.rs

```diff
@@ -1497,26 +1502,112 @@ impl Blockstore {
                 TransactionWithStatusMeta {
                     transaction: encoded_transaction,
                     meta: self
-                        .transaction_status_cf
-                        .get((slot, signature))
+                        .read_transaction_status((signature, slot))
                         .expect("Expect database get to succeed"),
                 }
             })
             .collect()
     }
 
+    fn initialize_transaction_status_index(&self) -> Result<()> {
+        self.transaction_status_index_cf
+            .put(0, &TransactionStatusIndexMeta::default())?;
+        self.transaction_status_index_cf
+            .put(1, &TransactionStatusIndexMeta::default())?;
+        // This dummy status improves compaction performance
+        self.transaction_status_cf.put(
+            (2, Signature::default(), 0),
+            &TransactionStatusMeta::default(),
+        )
+    }
+
+    fn toggle_transaction_status_index(
+        &self,
+        batch: &mut WriteBatch,
+        to_slot: Slot,
+    ) -> Result<Option<u64>> {
+        let index0 = self.transaction_status_index_cf.get(0)?;
+        if index0.is_none() {
+            return Ok(None);
+        }
+        let mut index0 = self.transaction_status_index_cf.get(0)?.unwrap();
+        let mut index1 = self.transaction_status_index_cf.get(1)?.unwrap();
+
+        if !index0.frozen && !index1.frozen {
+            index0.frozen = true;
+            batch.put::<cf::TransactionStatusIndex>(0, &index0)?;
+            Ok(None)
+        } else {
+            let result = if index0.frozen && to_slot >= index0.max_slot {
```

Comment:
I think `delete_range_cf(from, to)` is exclusive of `to`, so should this be `to_slot > index0.max_slot` (same comment for below)?

Reply:
Oh, good catch. I agree.

Reply:
Can we add a small test to catch corner cases like this?